hera_pspec/grouping.py

```diff
@@ -310,43 +362,227 @@ def average_spectra(uvp_in, blpair_groups=None, time_avg=False,
+                                for key in cov_real[0].keys():
+                                    cov_real[tindex][key] = cov_real[tindex][key]*w[tindex][None]
+                                    cov_imag[tindex][key] = cov_imag[tindex][key]*w[tindex][None]
+                                    cov_q_real[tindex][key] = cov_q_real[tindex][key]*w[tindex][None]
+                                    cov_q_imag[tindex][key] = cov_q_imag[tindex][key]*w[tindex][None]
+                                    var_real[tindex][key] = var_real[tindex][key]*w[tindex]
+                                    var_imag[tindex][key] = var_imag[tindex][key]*w[tindex]
+                                    var_q_real[tindex][key] = var_q_real[tindex][key]*w[tindex]
+                                    var_q_imag[tindex][key] = var_q_imag[tindex][key]*w[tindex]
+                            bpg_cov_real.append(cov_real)
+                            bpg_cov_imag.append(cov_imag)
+                            bpg_cov_q_real.append(cov_q_real)
+                            bpg_cov_q_imag.append(cov_q_imag)
+                            bpg_var_real.append(var_real)
+                            bpg_var_imag.append(var_imag)
+                            bpg_var_q_real.append(var_q_real)
+                            bpg_var_q_imag.append(var_q_imag)
 
 
                 # Take integration-weighted averages, with clipping to deal
                 # with zeros
                 bpg_data = np.sum(bpg_data, axis=0) \
                          / np.sum(w_list, axis=0).clip(1e-10, np.inf)
+                bpg_data_q = np.sum(bpg_data_q, axis=0) \
+                         / np.sum(w_list, axis=0).clip(1e-10, np.inf)
                 bpg_wgts = np.sum(bpg_wgts, axis=0) \
                          / np.sum(w_list, axis=0).clip(1e-10, np.inf)[:, None]
                 bpg_nsmp = np.sum(bpg_nsmp, axis=0)
                 bpg_ints = np.sum(bpg_ints, axis=0) \
                          / np.sum(w_list, axis=0).clip(1e-10, np.inf)
+                #if store_cov:
+                #    bpg_cov = np.sum(bpg_cov, axis=0) \
+                #            / np.sum(w_list, axis=0)[:,None].clip(1e-10, np.inf)
                 if store_cov:
-                    bpg_cov = np.sum(bpg_cov, axis=0) \
-                            / np.sum(w_list, axis=0)[:,None].clip(1e-10, np.inf)
+                    bpg_cov_real = np.array(bpg_cov_real)
+                    bpg_cov_imag = np.array(bpg_cov_imag)
+                    bpg_cov_q_real = np.array(bpg_cov_q_real)
+                    bpg_cov_q_imag = np.array(bpg_cov_q_imag)
+                    bpg_var_real = np.array(bpg_var_real)
+                    bpg_var_imag = np.array(bpg_var_imag)
+                    bpg_var_q_real = np.array(bpg_var_q_real)
+                    bpg_var_q_imag = np.array(bpg_var_q_imag)
+                    w_list = np.array(w_list)
+
+                    times = len(bpg_cov_real[0])
```

Comment:
The next hundred lines or so is an example where you can make an in-line function that does a general average, and then just feed it `bpg_cov_real`, or `bpg_cov_imag` etc. to do the operation on each array you want..

Reply:
I have tried on that. Will be more succinct in future commits. Thanks.